mne/realtime/base_client.py

```diff
@@ -0,0 +1,193 @@
+# Authors: Teon Brooks <teon.brooks@gmail.com>
+#          Mainak Jas <mainak@neuro.hut.fi>
+#
+# License: BSD (3-clause)
+
+import threading
+import time
+import numpy as np
+
+from ..utils import logger, fill_doc
+
+
+def _buffer_recv_worker(client):
+    """Worker thread that constantly receives buffers."""
+    try:
+        for raw_buffer in client.iter_raw_buffers():
+            client._push_raw_buffer(raw_buffer)
+    except RuntimeError as err:
+        # something is wrong, the server stopped (or something)
+        client._recv_thread = None
+        print('Buffer receive thread stopped: %s' % err)
+
+
+class _BaseClient(object):
+    """Base Realtime Client.
+
+    Parameters
+    ----------
+    identifier : str
+        The identifier of the server. IP address or LSL id or raw filename.
+    port : int | None
+        Port to use for the connection.
+    wait_max : float
+        Maximum time (in seconds) to wait for real-time buffer to start.
+    tmin : float | None
+        Time instant to start receiving buffers. If None, start from the latest
+        samples available.
+    tmax : float
+        Time instant to stop receiving buffers.
+    buffer_size : int
+        Size of each buffer in terms of number of samples.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see :func:`mne.verbose`
+        and :ref:`Logging documentation <tut_logging>` for more).
+    """
+
+    def __init__(self, identifier, port=None, wait_max=10., tmin=None,
+                 tmax=np.inf, buffer_size=1000, verbose=None):  # noqa: D102
+        self.identifier = identifier
+        self.port = port
+        self.wait_max = wait_max
+        self.tmin = tmin
+        self.tmax = tmax
+        self.buffer_size = buffer_size
+        self.verbose = verbose
+
+    def __enter__(self):  # noqa: D105
+
+        # connect to buffer
+        logger.info("Client: Waiting for server to start")
+        start_time = time.time()
+        while time.time() < (start_time + self.wait_max):
+            try:
+                self.connect()
+                logger.info("Client: Connected")
+                break
+            except Exception:
+                time.sleep(0.1)
+        else:
+            raise RuntimeError('Could not connect to Client.')
```

Comment:
are you going to refactor this in the next PR? :-)

Reply:
lol yup ;)